walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,108 @@
+  private Comparator<Event> eventComparator = new Comparator<Event>() {
+    @Override
+    public int compare(Event event1, Event event2) {
+      return event1.getWhen().start() - event2.getWhen().start();
+    }
+  };
+  /**
+   * Given a set of events and a meeting request, the query returns a list of
+   * free time ranges available for the meeting to be scheduled. If no valid
+   * time range is found, returns an empty list.
+   *
+   * Time complexity: 
+   * n: number of events
+   * m: number of attendees in every event
+   * O(nm)
+   *
+   * Space complexity:
+   * n: number of events
+   * m: number of available time ranges
+   * O(m+n)
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // We pass the events into a list since the collection can't be sorted,
+    // and we sort the list by event start time in ascending order.
+    List<Event> eventList = new ArrayList(events);
+    Collections.sort(eventList, eventComparator);
+
+    // Since there are no previous events yet, the start of day is also the
+    // start of the first range to check available time.
+    int pastEventEnd = TimeRange.START_OF_DAY;
+    int currentEventStart;
+    int availableTimeDuration;
+    List<TimeRange> availableTimes = new ArrayList<>();
+    for (Event event : eventList) {
+      if (!eventSharesAttendeesWithRequest(event, request)) {
+        // If the event does not share attendes with the request, it means that
+        // it does not change time availability, so it's skipped.
+        continue;
+      }
+      currentEventStart = event.getWhen().start();
+      availableTimeDuration = currentEventStart - pastEventEnd;
+
+      if (currentEventStart > pastEventEnd) {
+        // Example: |--A--|   |--B--|
+        if (availableTimeDuration >= request.getDuration()) {
+          availableTimes.add(TimeRange.fromStartDuration(
+              pastEventEnd, availableTimeDuration));
+        }
+        pastEventEnd = event.getWhen().end();
+      } 
+      else if (currentEventStart == pastEventEnd) {
+        // Example: |--A--|--B--|
+        pastEventEnd = event.getWhen().end();
+      }
+      else if (event.getWhen().end() > pastEventEnd) {
+        // Example: |--A--|
+        //              |--B--|
+        pastEventEnd = event.getWhen().end();
+      }
+
+      // There is no else, since would mean that the current event is
+      // completely absorbed by the previous event, meaning that the next
+      // available time still starts when the previous event ends.
+      // Example: |----A----|
+      //            |--B--|
+    }
+
+    // Now we just need to check the time from the end of the last meeting
+    // until end of day, basically doing the same procedure as before.
+    currentEventStart = TimeRange.END_OF_DAY;
+    if (currentEventStart - pastEventEnd >= request.getDuration()) {
```

Comment:
Perhaps just say `if (TimeRange.END_OF_DAY - pastEventEnd >= request.getDuration())...`, then you can keep currentEventStart as a local variable in your for loop.

Reply:
Done